### PR TITLE
📍 Add current location suggestion

### LINKED ISSUE #55: Search bar current location
## Description

Allow selecting current location as an option in the search bar/suggestions

## Todo

- When input is empty and user clicks on it, show Current Location as an option
- Don't show the option when the user starts typing
- Only show current location option if the user has allowed access
- Make a hook, useGeolocation, that returns the current location
- You need to use watchPosition in the hook so that the location actually updates
- Make a function in utils for reverse geocoding
- On desktop, add 1 or more UI components to allow for a button to be next to the input (see design)
- No current location in suggestions on desktop, only via button
- Put "Fixes #55" in the PR description

## Useful links

- https://github.com/streamich/react-use/blob/master/src/useGeolocation.ts
- https://getbootstrap.com/docs/4.0/components/input-group/ for inspiration
- Siraj

### PR DESCRIPTION
**Pull request by code-arman**
## Status:
:rocket: Ready

## Description
Adds current location as a suggestion on desktop and mobile. Goes to current location when the button or suggestion is clicked

Fixes #55 

## Screenshots

Screen Shot 2021-03-31 at 8 52 14 PM
Screen Shot 2021-03-31 at 8 53 44 PM

### CODE REVIEW

File: src/components/search/SearchEntry.js [hunk 1]
```diff
@@ -1,31 +1,38 @@
...
+  ({ children, isCurrent, ...props }, ref) => {
+    const [primaryText, secondaryText] = children
 
-  // TODO: Add current location to search dropdown
-  const leftIcon = (
-    <CircleIcon backgroundColor={theme.secondaryText}>
-      <Map />
-    </CircleIcon>
-  )
+    // TODO: Add current location to search dropdown
```
Comment (Qrtn) [inline src/components/search/SearchEntry.js:13]: Can you remove this TODO?

File: src/components/search/Search.js [hunk 2]
```diff
@@ -40,6 +67,9 @@ const StyledComboboxPopover = styled(ComboboxPopover)`
 
 const Search = (props) => {
   const { setViewport } = useContext(SearchContext)
+  const isDesktop = useIsDesktop()
+  const geoLocation = useGeolocation()
```
Comment (Qrtn) [inline src/components/search/Search.js:71]: `geolocation`

File: src/components/search/Search.js [hunk 3]
```diff
@@ -40,6 +67,9 @@ const StyledComboboxPopover = styled(ComboboxPopover)`
...
+  const [cityName, setCityName] = useState(undefined)
```
Comment (Qrtn) [inline src/components/search/Search.js:72]: `useState(null)`

File: src/components/search/Search.js [hunk 4]
```diff
@@ -56,30 +86,101 @@ const Search = (props) => {
...
   }
 
+  const inputRef = useRef(null)
+
+  useEffect(() => {
+    if (value === '') {
+      inputRef.current.blur()
+      inputRef.current.focus()
+    }
+  }, [value])
```
Comment (Qrtn) [inline src/components/search/Search.js:137]: Move these hooks to the beginning of function after the current "use"s. Rules of hooks

File: src/components/search/Search.js [hunk 5]
```diff
@@ -56,30 +86,101 @@ const Search = (props) => {
...
+
   return (
     <Combobox
       onSelect={handleSelect}
       aria-label="Search for a location"
+      openOnFocus={!isDesktop && cityName !== undefined}
       {...props}
     >
+      {console.log(cityName)}
```
Comment (Qrtn) [inline src/components/search/Search.js:146]: Remove this

File: src/components/search/Search.js [hunk 6]
```diff
@@ -56,30 +86,101 @@ const Search = (props) => {
...
+      inputRef.current.blur()
+      inputRef.current.focus()
+    }
+  }, [value])
+
   return (
     <Combobox
       onSelect={handleSelect}
       aria-label="Search for a location"
+      openOnFocus={!isDesktop && cityName !== undefined}
```
Comment (Qrtn) [inline src/components/search/Search.js:143]: `!isDesktop && cityName !== null`

File: src/utils/extractCityName.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+          )
+        } else {
+          reject('No results found')
+        }
+      } else {
+        reject(`Geocoder failed due to: ${status}`)
+      }
+    })
+  })
+}
```
Comment (Qrtn) [inline src/utils/extractCityName.js:19]: newline

File: src/components/search/Search.js [hunk 8]
```diff
@@ -89,13 +190,16 @@ const Search = (props) => {
               } = suggestion
 
               // Allow handleSelect to access the place id (see above)
-              descriptionToPlaceId.current[description] = place_id
+              if (place_id) {
+                descriptionToPlaceId.current[description] = place_id
+              }
```
Comment (Qrtn) [inline src/components/search/Search.js:195]: You don't need these if clause anymore right? You can move line 194 outside and remove the if?

File: src/components/search/Search.js [hunk 9]
```diff
@@ -56,30 +86,101 @@ const Search = (props) => {
...
+    if (description === 'Current Location') {
+      // Use fixed viewport around the lat and long of the current location
+      const lat = geoLocation.latitude
+      const lng = geoLocation.longitude
+
+      const viewportBounds = {
+        ne: { lat: lat + BOUND, lng: lng + BOUND },
+        sw: { lat: lat - BOUND, lng: lng - BOUND },
+      }
+      setViewport(viewportBounds)
```
Comment (Qrtn) [inline src/components/search/Search.js:121]: Why is this this line repeated in 127? Should there be an else clause here?

Comment (Qrtn) [reply src/components/search/Search.js:121]:
```
let viewportBounds;

if (description === current location) viewportBounds = ...
else viewportBounds = ...

setValue
setViewport
```

File: src/components/search/Search.js [hunk 10]
```diff
@@ -56,30 +86,101 @@ const Search = (props) => {
...
         as={Input}
         value={value}
         onChange={handleInput}
+        ref={inputRef}
         disabled={!ready}
         icon={<SearchAlt2 />}
+        prepend={
+          isDesktop && (
+            <StyledCurrentLocationButton
+              disabled={cityName === undefined}
```
Comment (Qrtn) [inline src/components/search/Search.js:157]: cityName == null

Comment (Qrtn) [reply src/components/search/Search.js:157]: Great work on styling the disabled version.

File: src/components/search/Search.js [hunk 11]
```diff
@@ -29,6 +35,27 @@ const getViewportBounds = async (placeId) => {
...
+  border-radius: 50% 0 0 50%;
+  border-right: 1px solid #e0e1e2;
+
+  svg {
+    color: ${({ theme }) => theme.blue};
+  }
+
+  &:disabled svg {
+    color: ${({ theme }) => theme.tertiaryText};
+  }
```
Comment (Qrtn) [inline src/components/search/Search.js:57]: Good work with the prepend and append!

However, you shouldn't have to style the border of the prepend. That should go in the Input. Think about it. If you were using the Input component, you shouldn't be expected to style the border-right, radius, etc. You should only have to think about styling the thing that you're putting inside the prepend or append of the Input.

File: src/components/search/Search.js [hunk 12]
```diff
@@ -8,15 +8,21 @@ import {
...
 // TODO: Switch to https://www.npmjs.com/package/@googlemaps/js-api-loader
 import usePlacesAutocomplete, { getGeocode } from 'use-places-autocomplete'
 
+import extractCityName from '../../utils/extractCityName'
+import { useIsDesktop } from '../../utils/useBreakpoint'
 import Input from '../ui/Input'
 import SearchContext from './SearchContext'
 import SearchEntry from './SearchEntry'
 
+const BOUND = 0.001
```
Comment (Qrtn) [inline src/components/search/Search.js:24]: Can you be more specific with this name? `CURRENT_LOCATION_VIEWPORT_RADIUS`

File: src/components/search/DesktopSearch.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
+import { CurrentLocation } from '@styled-icons/boxicons-regular/CurrentLocation'
```
Comment (Qrtn) [inline src/components/search/DesktopSearch.js:1]: Can you remove this file?

**Review by Qrtn — CHANGES_REQUESTED** [review]
Really great work overall. Main thing that needs to be changed is considering what styles need to go where. People passing stuff into prepend or append of Input shouldn't have to worry about how to style what's around the stuff they're passing in (the border, etc.) Also make sure to pull from main again.

File: src/components/search/Search.js [hunk 14]
```diff
@@ -40,48 +60,128 @@ const StyledComboboxPopover = styled(ComboboxPopover)`
...
+      viewportBounds = {
+        ne: {
+          lat: latitude + CURRENT_LOCATION_VIEWPORT_RADIUS,
+          lng: longitude + CURRENT_LOCATION_VIEWPORT_RADIUS,
+        },
+        sw: {
+          lat: latitude - CURRENT_LOCATION_VIEWPORT_RADIUS,
+          lng: longitude - CURRENT_LOCATION_VIEWPORT_RADIUS,
+        },
+      }
```
Comment (Qrtn) [inline src/components/search/Search.js:130]: I actually think it'd help readability to make a function that takes geolocation and returns bounds around it. `getGeolocationViewportBounds`

File: src/components/ui/Input.js [hunk 15]
```diff
@@ -1,9 +1,38 @@
...
+  }
+  height: 100%;
+  & > * {
+    background: none;
+    color: inherit;
+    border: none;
+    height: 100%;
+    border-radius: ${(props) => (props.prepend ? '50% 0 0 50%' : 'none')};
+    border-right: ${(props) => (props.prepend ? '1px solid #e0e1e2' : 'none')};
+    border-left: ${(props) => (!props.prepend ? '1px solid #e0e1e2' : 'none')};
```
Comment (Qrtn) [inline src/components/ui/Input.js:18]: Use the theme color here instead of `#e0e1e2`

Comment (Qrtn) [reply src/components/ui/Input.js:18]: @SirajChokshi

File: src/components/ui/Input.js [hunk 16]
```diff
@@ -59,6 +91,7 @@ const StyledInput = styled(Input)`
       width: 100%;
       outline: none;
       height: 44px;
+      padding-left: ${(props) => (props.prepend ? '10px' : 0)};
```
Comment (Qrtn) [inline src/components/ui/Input.js:94]: Do you have the equivalent right padding for append? How does append look?

**Review by Qrtn — APPROVED** [review]
Great work Arman, just a few more comments.

File: src/components/ui/Input.js (hunk 16 above)
Comment (code-arman) [reply src/components/ui/Input.js:94]: That line adds space between the gray bar and the "Search for..." text
Screen Shot 2021-04-01 at 5 25 21 PM

This is what it looks like with no padding for the append (how the code currently is)
Screen Shot 2021-04-01 at 5 24 11 PM
If I add the 10px of padding to the append, it looks like this (too much space between left outline and text)

Screen Shot 2021-04-01 at 5 24 47 PM


Could be wrong tho

File: src/components/search/SearchEntry.js [hunk 17]
```diff
@@ -1,31 +1,37 @@
...
-      {...props}
-    />
-  )
-})
+    return (
+      <ListEntry
+        leftIcons={leftIcon}
+        primaryText={primaryText}
+        secondaryText={secondaryText}
+        rightIcons={<Navigation size={20} color={'#4183C4'} />}
```
Comment (n3a9) [inline src/components/search/SearchEntry.js:28]: hardcoded!

**Review by n3a9 — APPROVED** [review]
eh should be fine if jeffrey says its ok

**Comment by code-arman** [comment]
> eh should be fine if jeffrey says its ok

get off this repo cuh

File: src/components/desktop/SidePane.js [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import styled from 'styled-components/macro'
+
+import Search from '../search/Search'
+
+const PaneContainer = styled.div`
+  padding: 20px 10px;
+  padding-right: 0px;
```
Comment (SirajChokshi) [inline src/components/desktop/SidePane.js:7]: We can make this one rule with `padding: 20px 0 20px 10px;`

File: src/components/desktop/SidePane.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
+import styled from 'styled-components/macro'
+
+import Search from '../search/Search'
+
+const PaneContainer = styled.div`
```
Comment (SirajChokshi) [inline src/components/desktop/SidePane.js:5]: Do we need a while styled component for this? Its just a wrapper, but maybe its the cleaner solution.

File: src/components/entry/EntryDetails.js [hunk 20]
```diff
@@ -143,7 +144,7 @@ const EntryDetails = ({ isDesktop }) => {
         <TagList>
           {locationData.access && <Tag>{ACCESS_TYPE[locationData.access]}</Tag>}
           {/* TODO: Siraj - Put tag colors in theme/use constants somehow/map from object */}
-          <Tag color="#4183C4" backgroundColor="#D9E6F3">
+          <Tag color={theme.blue} backgroundColor="#D9E6F3">
```
Comment (SirajChokshi) [inline src/components/entry/EntryDetails.js:147]: Could we make the background color a theme value as well?

File: src/components/search/locationInfo.js [hunk 21]
```diff
@@ -0,0 +1,38 @@
+const resultToLocationInfo = (result) => {
```
Comment (SirajChokshi) [inline src/components/search/locationInfo.js:1]: Maybe we could add some comments here so we know what these helpers do?

**Review by SirajChokshi — APPROVED** [review]
Good work! Just a few nits 😎

File: src/components/search/locationInfo.js (hunk 21 above)
Comment (Qrtn) [reply src/components/search/locationInfo.js:5]: Done!

**Pull request merged**